daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -483,6 +483,132 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
+          case _ => Right(copy(keyInputs = keyInputs.updated(key, value)))
+        }
+      }
+      def assertKeyMapping(
+          templateId: Identifier,
+          cid: Value.ContractId,
+          optKey: Option[Node.KeyWithMaintainers[Value[Value.ContractId]]],
+      ): Either[GlobalKey, State] =
+        optKey.fold[Either[GlobalKey, State]](Right(this)) { key =>
+          val gk = GlobalKey.assertBuild(templateId, key.key)
+          keys.get(gk) match {
+            case Some(keyMapping) if Some(cid) != keyMapping => Left(gk)
+            case _ =>
+              val r = copy(keys = keys.updated(gk, Some(cid)))
+              if (localContracts.contains(cid)) {
+                Right(r)
+              } else {
+                r.setKeyMapping(gk, Some(cid))
+              }
+          }
+        }
+      def handleExercise(exe: Node.NodeExercises[_, Value.ContractId]) =
+        assertKeyMapping(exe.templateId, exe.targetCoid, exe.key).map { state =>
+          exe.key.fold(state) { key =>
+            val gk = GlobalKey.assertBuild(exe.templateId, key.key)
+            if (exe.consuming) {
+              state.copy(
+                keys = keys.updated(gk, None)
+              )
+            } else {
+              state
+            }
+          }
+        }
+
+      def handleCreate(create: Node.NodeCreate[Value.ContractId]) =
+        create.key.fold[Either[GlobalKey, State]](Right(this)) { key =>
+          val gk = GlobalKey.assertBuild(create.templateId, key.key)
+          keys.get(gk).orElse(keyInputs.get(gk)) match {
+            case None =>
+              Right(
+                copy(
+                  keys = keys.updated(gk, Some(create.coid)),
+                  keyInputs = keyInputs.updated(gk, None),
+                )
+              )
+            case Some(None) =>
+              Right(copy(keys = keys.updated(gk, Some(create.coid))))
+            case Some(Some(_)) =>
+              Left(gk)
+          }
+        }
+
+      def handleLookup(lookup: Node.NodeLookupByKey[Value.ContractId]) = {
+        val gk = GlobalKey.assertBuild(lookup.templateId, lookup.key.key)
+        keys.get(gk) match {
+          case None => setKeyMapping(gk, lookup.result)
+          case Some(optCid) =>
+            if (optCid != lookup.result) {
+              Left(gk)
+            } else {
+              // No need to update anything, we updated keyInputs when we updated keys.
+              Right(this)
+            }
+        }
+      }
+
+      def handleLeaf(leaf: Node.LeafOnlyActionNode[Value.ContractId]): Either[GlobalKey, State] =
+        leaf match {
+          case create: Node.NodeCreate[Value.ContractId] =>
+            handleCreate(create)
+          case fetch: Node.NodeFetch[Value.ContractId] =>
+            assertKeyMapping(fetch.templateId, fetch.coid, fetch.key)
+          case lookup: Node.NodeLookupByKey[Value.ContractId] =>
+            handleLookup(lookup)
+        }
+      def beginRollback(): State =
```

Comment:
```suggestion
      def beginRollback: State =
```

Reply:
Just to improve my scala knowledge. Is there a good reason why you prefer this? In Scala 2.13, there is a difference between the two and the call below will fail unless I remove the parentheses (and the other way around).

Reply:
Thanks!